Common/Statistics/TradeBuilder.cs

```diff
@@ -0,0 +1,486 @@
+{
+    /// <summary>
+    /// The TradeBuilder class generates trades from executions and market price updates
+    /// </summary>
+    public class TradeBuilder
+    {
+        /// <summary>
+        /// Helper class to manage pending trades and market price updates for a symbol
+        /// </summary>
+        private class Position 
+        {
+            internal List<Trade> PendingTrades { get; set; }
+            internal List<OrderEvent> PendingFills { get; set; }
+            public decimal TotalFees { get; set; }
+            internal decimal MaxPrice { get; set; }
+            internal decimal MinPrice { get; set; }
+
+            public Position()
+            {
+                PendingTrades = new List<Trade>();
+                PendingFills = new List<OrderEvent>();
+            }
+        }
+
+        private readonly List<Trade> _closedTrades = new List<Trade>();
+        private readonly Dictionary<string, Position> _positions = new Dictionary<string, Position>();
+        private readonly HashSet<int> _ordersWithFeesAssigned = new HashSet<int>();
+        private readonly FillGroupingMethod _groupingMethod;
+        private readonly FillMatchingMethod _matchingMethod;
+        private bool _liveMode;
+
+        private const int LiveModeMaxTradeCount = 10000;
+        private const int LiveModeMaxTradeAgeMonths = 12;
+
+        /// <summary>
+        /// Initializes a new instance of the TradeBuilder class
+        /// </summary>
+        public TradeBuilder(FillGroupingMethod groupingMethod, FillMatchingMethod matchingMethod)
+        {
+            _groupingMethod = groupingMethod;
+            _matchingMethod = matchingMethod;
+        }
+
+        /// <summary>
+        /// Sets the live mode flag
+        /// </summary>
+        /// <param name="live">The live mode flag</param>
+        public void SetLiveMode(bool live)
+        {
+            _liveMode = live;
+        }
+
+        /// <summary>
+        /// The list of closed trades
+        /// </summary>
+        public List<Trade> ClosedTrades
+        {
+            get { return _closedTrades; }
+        }
+
+        /// <summary>
+        /// Returns true if there is an open position for the symbol
+        /// </summary>
+        /// <param name="symbol">The symbol</param>
+        /// <returns>true if there is an open position for the symbol</returns>
+        public bool HasOpenPosition(string symbol)
+        {
+            Position position;
+            if (!_positions.TryGetValue(symbol, out position)) return false;
+
+            if (_groupingMethod == FillGroupingMethod.FillToFill)
+                return position.PendingTrades.Count > 0;
+
+            return position.PendingFills.Count > 0;
+        }
+
+        /// <summary>
+        /// Sets the current market price for the symbol
+        /// </summary>
+        /// <param name="symbol"></param>
+        /// <param name="price"></param>
+        public void SetMarketPrice(string symbol, decimal price)
```

Comment:
Agreed on the extra dependency, when all is said and done we'll need to do some performance testing to confirm/deny the impact.